### ISSUE TITLE
Remove GCC_HOST_COMPILER_PREFIX as it may be out of sync with GCC_HOST_COMPILER_PATH

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#39263]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): RHEL 7.5
- TensorFlow installed from (source or binary): source
- TensorFlow version: 2.2.0
- Python version: 3.7.4
- Bazel version (if compiling from source): 2.0.0
- GCC/Compiler version (if compiling from source): 7.3.0

**Describe the problem**

The file https://github.com/tensorflow/tensorflow/blob/1588f45ee56860d247a1c26ea228cb3721b4bf1b/third_party/gpus/cuda_configure.bzl has a documented environment variable `GCC_HOST_COMPILER_PATH` to set the path (or name) of a GCC host compiler. However it also uses an undocumented variable `GCC_HOST_COMPILER_PREFIX` to get the folder where the gcc binary resides (guessing from name) defaulting to `/usr/bin` if it isn't set:
https://github.com/tensorflow/tensorflow/blob/1588f45ee56860d247a1c26ea228cb3721b4bf1b/third_party/gpus/cuda_configure.bzl#L1028-L1030

I have 2 problems with that:
- It is undocumented and hence hard to set right if you don't know for sure what it is
- The default is wrong when the (documented!) `GCC_HOST_COMPILER_PATH`  is used

This leads to issues such as 

```
 external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc -o bazel-out/k8-opt/bin/external/protobuf_archive/js_embed -Wl,-no-as-needed -pie -Wl,-z,relro,-z,now '-Wl,--build-id=md5' '-Wl,--hash-style=gnu' -no-canonical-prefixes -B/usr/bin -Wl,--gc-sections -Wl,@bazel-out/k8-opt/bin/external/protobuf_archive/js_embed-2.params)
/usr/bin/ld.gold: error: /software/software/GCCcore/7.3.0/lib/gcc/x86_64-pc-linux-gnu/7.3.0/crtbeginS.o: unsupported reloc 42 against global symbol _ITM_deregisterTMCloneTable
/usr/bin/ld.gold: error: /software/software/GCCcore/7.3.0/lib/gcc/x86_64-pc-linux-gnu/7.3.0/crtbeginS.o: unsupported reloc 42 against global symbol _ITM_registerTMCloneTable
/usr/bin/ld.gold: error: bazel-out/k8-opt/bin/external/protobuf_archive/_objs/js_embed/embed.o: unsupported reloc 42 against global symbol std::ios_base::Init::~Init()
/software/software/GCCcore/7.3.0/lib/gcc/x86_64-pc-linux-gnu/7.3.0/crtbeginS.o(.text+0x1a): error: unsupported reloc 42
/software/software/GCCcore/7.3.0/lib/gcc/x86_64-pc-linux-gnu/7.3.0/crtbeginS.o(.text+0x6b): error: unsupported reloc 42
bazel-out/k8-opt/bin/external/protobuf_archive/_objs/js_embed/embed.o:embed.cc:function _GLOBAL__sub_I_main: error: unsupported reloc 42
collect2: error: ld returned 1 exit status
```

As reported at https://github.com/easybuilders/easybuild-easyconfigs/pull/7800#issuecomment-471447493

I hence propose to either completely remove that variable in favor of deriving its value from `GCC_HOST_COMPILER_PATH` or properly documenting it with a better default.

It does not seem to be required at all so it's likely best to just remove it. This should have been done by #34218 but for some reason that merge was reverted with a very misleading commit title: https://github.com/tensorflow/tensorflow/commit/f0571998d0195b5b243cf409a64d5fa17bd44d43

@gunan @mihaimaruseac please take a look what went wrong

**Comment by r4nt** [comment]
Before that flag was introduced we hard-coded -B/usr/bin, which is arguably significantly worse :)
-B/usr/bin was needed because currently the toolchain is not good at figuring out where other binutils are when they are not in the same path as the compiler, which is a not uncommon setup for people building their own compilers.

**Comment by r4nt** [comment]
The right solution is to actually discover all binutils via PATH, like other build systems would.

**Comment by Flamefire** [comment]
From the comment:

>     # TODO: when bazel stops adding '-B/usr/bin' by default, remove this
    #       flag from the CROSSTOOL completely (see
    #       https://github.com/bazelbuild/bazel/issues/563

And as explained in #34202:

> As the underlying Bazel issue bazelbuild/bazel#5634 is resolved, this code can (and should) go now

So I don't see why setting this is still required. And I also disagree with "Before that flag was introduced": That flag was never documented as far as I can tell. So are people supposed to use/rely on it?

For another datapoint: In Easybuild (install automation tool) we patch that file to set this to empty since forever as adding `/usr/bin` leads to (the mentioned) problems and haven't had any issues.

**Comment by r4nt** [comment]
We're currently using this in our cross-compilation setup. We use a cross-compiler gcc for manylinux, but the host system's binutils. Removing this will get rolled back until somebody has time to hunt down regressions, root-cause them and then figure out how we'll need to adapt our setup. I think you're right that it can be removed, but it's a bit of work to change & test everything that relies on it.

Re: the commit message: what's wrong about it?

**Comment by r4nt** [comment]
It also looks like there are still open issues in bazel around this:
https://github.com/bazelbuild/bazel/issues/6834

**Comment by Flamefire** [comment]
> We're currently using this in our cross-compilation setup.

Can you clarify what exactly you mean by "this"?   
Especially did you try with #34218 applied (i.e. `cuda_defines["%{linker_bin_path}"] = ""`)? I'd expect not setting the path would make it pick the binutils from PATH.

Ok just read the linked bazel issue and it seems PATH is discarded by Bazel.

Then the comment quoted above (`when bazel stops adding '-B/usr/bin' by default, remove this`) is clearly wrong. This results in my second alternative of the proposal: "properly documenting it with a better default."

So at the very least it should be documented because as the linked commit from bazelbuild/bazel#6834 states:
> and [adding /usr/bin] was done to get Tensor Flow building on some (but failing on other?) versions of RedHat

So there are known failures with that added as well as without it. I'm wondering why it is now only done for the CUDA toolchain though and not for anything else.

> Re: the commit message: what's wrong about it?

The commit message is "Merge pull request #34218 from Flamefire:fix_missing_linker_path" but it does revert that merge. Also interesting: The merge commit has a green CI tick while the revert commit has a failed CI tick :shrug: 

So summary:
- Check if that `linker_bin_path` setting is still required and remove if not. Otherwise:
- Document `GCC_HOST_COMPILER_PREFIX` like the other variables
- Opt (but recommended) default to the path of the `GCC_HOST_COMPILER`, maybe checking if binutils can be found there. IMO this is the right thing because having binutils separate is the exception rather than the norm.

**Comment by Flamefire** [comment]
Yes, the comment from https://github.com/tensorflow/tensorflow/issues/39263#issuecomment-627207228 still applies as-is with latest master. See the summary there

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by Flamefire** [comment]
I still suggest to rename `GCC_HOST_COMPILER_PREFIX` to e.g.  `HOST_BINUTILS_PATH` and document it e.g. in `third_party/gpus/cuda_configure.bzl` where a section reads

> `cuda_configure` depends on the following environment variables:

Also `configure.py` could set it, which needs to query and/or check if that path makes sense, e.g. `ar` is found there

**Comment by Venkat6871** [comment]
Hi @Flamefire ,
Thank you so much for filing this build/install request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!

**Comment by Flamefire** [comment]
This needs a decision on how to handle this that does still work for everyone, i.e. an answer to the questions and suggestions above.
